DBZ-7862 Updated MongoDB connector docs to reflect replica_set connection mode removal (#5556)
debezium/debezium: documentation/modules/ROOT/pages/connectors/mongodb.adoc

`capture.scope` is set to `database`::
Grant the user permission to read the database specified by the connector's xref:mongodb-property-capture-target[`capture.target`] property.

.Permission to use the MongoDB `hello` command

Regardless of the `capture.scope` setting, the user requires permission to run the MongoDB https://www.mongodb.com/docs/manual/reference/command/hello/[hello] command.

.Permission to read the `config.shards` collection
Depending on your {prodname} environment, to enable the connector to perform offset consolidation, you must grant the connector user explicit permission to read the `config.shards` collection.  
Permission to read the `config.shards` collection is required for the following connector environments:

> * The connector was upgraded from a {prodname} 2.5 or earlier.
> * The connector is configured to capture changes from a sharded MongoDB cluster.

Reviewer: ```suggestion
* Connectors upgraded from {prodname} 2.5 or earlier.
* Connectors configured to capture changes from a sharded MongoDB cluster.
```

You should give each MongoDB connector a unique logical name that meaningfully describes the source MongoDB system.
We recommend logical names begin with an alphabetic or underscore character, and remaining characters that are alphanumeric or underscore.

// Type: concept
// Title: How {prodname} MongoDB connectors perform offset consolidation
// ModuleID: how-debezium-mongodb-connectors-perform-offset-consolidation
[[mongodb-offset-consolidation]]
=== Offset consolidation

> Version 2.6.0 of Debezium MongoDb connector removed the `replica_set` way of connecting to a sharded MongoDb deployment. Consequently, the offsets  recorded by a previous version of MongoDb connector which used the `replica_set` connection mode are incompatible with MongoDb connector version 2.6.0 and later.

Author: ok

> To minimise the impact of this change as well as to prevent unexpected snapshot re-execution the Debezium MongoDb connector version 2.6.0 or later will perform what we call an offset consolidation procedure to determine the previously recorded offset:

Reviewer: Not sure if you already caught it, but I just noticed that I introduced a duplicate word ("to to .."), so I removed it in the updated edit.

1. If an offset recorded by MongoDb connector 2.6.0 or later exists, it will be used
2. If a compatible offset recorded by older MongoDb connector version exists, it will be used
    - The offset will be compatible if the  targeted MongoDb was deployed as a replica set, or if the `sharded` connection mode was used by the connector to capture changes from sharded MongoDb deployment.
3. If shard specific offsets recorded by older MongoDB connector version exit for all current database shards.
    a. If offset invalidation is xref:mongodb-property-allow-offset-invalidation[allowed], the oldest of shard specific offsets will be used.
    b. If offset invalidation is xref:mongodb-property-allow-offset-invalidation[not allowed],the connector will fail to start

> 4. No offset was previously recorded.

Author: No, it concludes that no previous offset exists and starts with an empty one

> 4. No offset was previously recorded.

Reviewer: > No, it concludes that no previous offset exists and starts with an empty one

OK. Assuming that your comment relates to the statement about committing previously unrecorded offsets, I've revised that line, and added it back as Step 4 in the procedure.

> 4. No offset was previously recorded.

Author: Better. Do you think it's also worth stating that if no existing offsets were found the connector will perform snapshot as usual?

> 4. No offset was previously recorded.

Reviewer: > Better. Do you think it's also worth stating that if no existing offsets were found the connector will perform snapshot as usual?

Good point, thanks. Updated.

You should give each MongoDB connector a unique logical name that meaningfully describes the source MongoDB system.
We recommend logical names begin with an alphabetic or underscore character, and remaining characters that are alphanumeric or underscore.

// Type: concept
// Title: How {prodname} MongoDB connectors perform offset consolidation
// ModuleID: how-debezium-mongodb-connectors-perform-offset-consolidation
[[mongodb-offset-consolidation]]
=== Offset consolidation

> The {prodname} MongoDB connector no longer supports `replica_set` connections to sharded MongoDB deployments. 

Reviewer: ```suggestion

The {prodname} MongoDB connector no longer supports `replica_set` connections to sharded MongoDB deployments. 
```
Insert a blank line after the heading to prevent intermittent problems in the downstream dos in which paragraphs the immediately follow headings are swallowed up and fail to render.

As a result, the offsets recorded by connector versions that used the `replica_set` connection mode are incompatible with the current version.

> To minimize the way that the connection mode change affects the connector, and to prevent the connector from running unnecessary snapshots, after you upgrade the {prodname} MongoDB connector, it runs an offset consolidation procedure.

Reviewer: ```suggestion
To minimize the affects of the connection mode change, and to prevent the connector from running unnecessary snapshots, when the connector restarts after the upgrade, it runs a procedure to consolidate offsets.
```

> During the consolidation procedure, the connector runs through the following steps to  to reconcile offsets that were recorded by the earlier connector version:

Reviewer: ```suggestion
During this offset consolidation procedure, the connector completes the following steps to reconcile offsets that were recorded by the earlier connector version:
```

1. Offsets that were recorded by connector versions later than 2.5 are used as-is.
2. Offsets for events that were captured in `sharded` connection mode from sharded MongoDB deployments, or from MongoDB replica set deployments, are used as-is.
3. Shard-specific offsets that were recorded by connector versions 2.5.x and earlier are used as-is, if both of the following conditions apply:

* The offsets exist for all current database shards.
* xref:mongodb-property-allow-offset-invalidation[Offset invalidation] is enabled. +
If offset invalidation is disabled, the connector fails to start.

> 4.  After the connector completes the preceding steps to process existing offsets, it resumes streaming from the collection, and then commits offsets for new events that it captures. +

Reviewer: ```suggestion
4. After the connector processes existing offsets in the preceding steps, it resumes streaming from the collection, and then commits offsets for new events that it captures. +
```

Author: @roldanbob  Good thing you did the pass, I would have missed this. The statement "it resumes streaming from the collection" doesn't really make sense for multiple reasons 

- Changes are technically always streamed from a change stream not from  mongo/db/collection
- Changes stream can be opened against Mongo instance/database/collection
- We actually don't support changes streams scoped to collection yet 

In this case I would use generic "... it resumes streaming changes, and then..."

|===
|Property |Default |Description

|[[mongodb-property-allow-offset-invalidation]]<<mongodb-property-allow-offset-invalidation, `+internal.mongodb.allow.offset.invalidation+`>>
|false

> |Allows shard specific offsets recorded by older connector versions to be invalidated and xref:mongodb-offset-consolidation[consolidated]

Reviewer: ```suggestion
|Set this property to `true` to enable the connector to invalidate and xref:mongodb-offset-consolidation[consolidate] shard-specific offsets that were recorded by earlier connector versions.
```

[WARNING]
====

> This property is considered internal, it will be removed and the default behaviour will be as if enabled in future releases.

Reviewer: ```suggestion
This is an internal property, and is scheduled for removal.

ifdef::community[]
Subsequent releases of the connector will invalidate and consolidate shard-specific offsets.  
endif::community[]
```
For the downstream docs, I think that it's ok to "pre-announce" removal of the property, but we should not make statements that describe possible future behavior.

Author: Lets hear @jpechane here as this is somwewhat special case. This property was introduced as temporary. It's purpose is only for users to transition from one model to the other.

Reviewer 2: I think the text should go to the downstream docs as well. If @roldanbob is not fully on-board wit definitive statement I think it could be qualified like "It is expected that subsequent releases" or something like that signals high probability but not 100 % certainity.

Reviewer: OK, I think that I've come up with wording that might satisfy. WDYT?

|===
|Property |Default |Description

|[[mongodb-property-allow-offset-invalidation]]<<mongodb-property-allow-offset-invalidation, `+internal.mongodb.allow.offset.invalidation+`>>
|false
|Set this property to `true` to enable the connector to invalidate and xref:mongodb-offset-consolidation[consolidate] shard-specific offsets that were recorded by earlier connector versions.

[WARNING]
====

This property permits you to modify the current default behavior.

> It is subject to removal if a future release changes the default behavior so that a connector automatically invalidates and consolidates offsets that are recorded by earlier connector versions.

Reviewer: ```suggestion
The property is subject to removal in a future release if the default behavior changes to permit the connector to automatically invalidate and consolidate offsets that are recorded by earlier connector versions.
```
